runtime/lua/vim/lsp/util.lua

```diff
@@ -683,6 +684,52 @@ do
       api.nvim_buf_set_virtual_text(bufnr, diagnostic_ns, line, virt_texts, {})
     end
   end
+  function M.buf_diagnostics_statusline(bufnr, diagnostics)
+    all_buffer_diagnostics_counts[bufnr] = { errors=0, warnings=0, hints=0, info=0 }
+    for _, diagnostic in ipairs(diagnostics) do
+      if diagnostic.severity == 1 then
+        all_buffer_diagnostics_counts[bufnr]['errors'] = all_buffer_diagnostics_counts[bufnr]['errors'] + 1
+      elseif diagnostic.severity == 2 then
+        all_buffer_diagnostics_counts[bufnr]['warnings'] = all_buffer_diagnostics_counts[bufnr]['warnings'] + 1
+      elseif diagnostic.severity == 3 then
+        all_buffer_diagnostics_counts[bufnr]['info'] = all_buffer_diagnostics_counts[bufnr]['info'] + 1
+      elseif diagnostic.severity == 4 then
+        all_buffer_diagnostics_counts[bufnr]['hints'] = all_buffer_diagnostics_counts[bufnr]['hints'] + 1
+      end
+    end
+  end
+  function M.buf_diagnostics_error_count()
+      local bufnr = vim.api.nvim_get_current_buf()
+      if all_buffer_diagnostics_counts[bufnr] ~= nil then
+          return all_buffer_diagnostics_counts[bufnr]['errors']
+      else
+          return '-'
+      end
+  end
+  function M.buf_diagnostics_warning_count()
```

Comment:
4 new functions is far too many. It increases the API size for little gain. Each function needs to be documented, etc. Seems pretty natural to have 1 function that takes a parameter instead, not to mention that would eliminate 20 lines of redundant code in this PR.

Reply:
makes total sense, will change